File: pm/api/views/role.py
from typing import TYPE_CHECKING, Self

from beanie import PydanticObjectId
from pydantic import BaseModel

import pm.models as m
from pm.permissions import Permissions

__all__ = (
    'RoleOutput',
    'RoleLinkOutput',
    'PERMISSIONS_BY_CATEGORY',
)


PERMISSIONS_BY_CATEGORY = {
    'Project': {
        Permissions.PROJECT_READ: 'Read project',
        Permissions.PROJECT_UPDATE: 'Update project',
        Permissions.PROJECT_DELETE: 'Delete project',
    },
    'Issue': {
        Permissions.ISSUE_CREATE: 'Create issue',
        Permissions.ISSUE_READ: 'Read issue',
        Permissions.ISSUE_UPDATE: 'Update issue',
        Permissions.ISSUE_DELETE: 'Delete issue',
    },
    'Comment': {
        Permissions.COMMENT_CREATE: 'Create comment',
        Permissions.COMMENT_READ: 'Read comment',
        Permissions.COMMENT_UPDATE: 'Update comment',
        Permissions.COMMENT_DELETE_OWN: 'Delete own comment',
        Permissions.COMMENT_DELETE: 'Delete comment',
    },
}


class PermissionOutput(BaseModel):
    key: Permissions
    label: str
    granted: bool


class PermissionCategoryOutput(BaseModel):
    label: str
    permissions: list[PermissionOutput]


class RoleLinkOutput(BaseModel):
    id: PydanticObjectId
    name: str
    description: str | None

    @classmethod
    def from_obj(cls, obj: m.Role | m.RoleLinkField) -> Self:
        return cls(
            id=obj.id,
            name=obj.name,
            description=obj.description,
        )


class RoleOutput(BaseModel):
    id: PydanticObjectId
    name: str
    description: str | None
    permissions: list[PermissionCategoryOutput]

    @classmethod
    def from_obj(cls, obj: m.Role | m.RoleLinkField) -> Self:
        role_permissions = set(obj.permissions)
        return cls(
            id=obj.id,
            name=obj.name,
            description=obj.description,
            permissions=[
                PermissionCategoryOutput(
                    label=category,
                    permissions=[
                        PermissionOutput(
                            key=key,
                            label=label,
                            granted=key in role_permissions,
                        )
                        for key, label in permissions.items()
                    ],
                )
                for category, permissions in PERMISSIONS_BY_CATEGORY.items()
            ],
        )

File: frontend/src/types/role.ts
export type CreateRoleT = {
    name: string;
    description: string | null;
};

export type UpdateRoleT = Partial<CreateRoleT>;

export const permissionKeys = [
    "project:read",
    "project:update",
    "project:delete",
    "issue:create",
    "issue:read",
    "issue:update",
    "issue:delete",
    "comment:create",
    "comment:read",
    "comment:update",
    "comment:delete_own",
    "comment:delete",
    "attachment:create",
    "attachment:read",
    "attachment:delete_own",
    "attachment:delete",
] as const;

export type PermissionKeyT = (typeof permissionKeys)[number];

export type RolePermissionT = {
    key: PermissionKeyT;
    label: string;
    granted: boolean;
};

export type PermissionGroupT = {
    label: string;
    permissions: RolePermissionT[];
};

export type RoleT = {
    id: string;
    name: string;
    description: string;
    permissions: PermissionGroupT[];
};
